Add Task Submission / Return Hooks

cswartzvi opened the pull request:
Current pre/post task execution hooks may run on out-of-process executors potentially causing issues with certain logging and stateful adapters. This PR adds additional hooks that are run *before* a task is submitted to an executor and *after* a task future is resolved - both on the main process.

## Changes

I added task submission / resolution hooks (`BasePreTaskSubmission` and `BasePostTaskResolution`) and adapters (`TaskSubmissionHook` and `TaskResolutionHook`) to the codebase as well as updated the main task execution function `run_graph_to_completion` to call the new hooks. In support of this I made `TaskImplementation` hashable via it's `task_id` in order to avoid a parallel structure - @elijahbenizzy perhaps in the future this could be used to a queue-based task processor?

## How I tested this

I added individual tests for both `TaskSubmissionHook` and `TaskResolutionHook` and added them to `test_multi_hook`.

## Notes
1. The docs for `TaskSubmissionHook` and `TaskResolutionHook`, as well some other task execution hooks, have been updated. I think we could add some more details - maybe a flow chart of when the individual hooks are called? I can add that if the current changes seems reasonable.
2. `TaskResolutionHook` currently has an optional  `error` attribute. I am on the fence as to whether or not this is needed. My thinking is that since `TaskFuture` essentially wraps `concurrent.futures.Future` we could forward the `exception` call. Thoughts? Currently `error`  is always `None`.

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

cswartzvi commented:
Ahh - used `str | None`

cswartzvi commented:
Regarding the failing doc test - possibly related to `ddtrace` 3.1.0 update, in particular https://github.com/DataDog/dd-trace-py/pull/12186

skrawcz commented:
What's the impact here to the Hamilton Tracker? Should we implement these new functions for it?

cswartzvi commented:
@skrawcz,

I am not that familiar with the `HamiltonTracker`, but it appears to implement, mainly, node-based hooks. This change should not alter their function - node-based hooks should run as before. 

Offhand, do you know if the UI has any issues tracking out-of-process execution? I see that `HamiltonTracker` is stateful and I don't see any evident `__getstate__` or `__setstate__` methods, but I could be missing something. If there are issues, these hooks might be of some use - I would need to experiment.

skrawcz commented:
> I am not that familiar with the `HamiltonTracker`, but it appears to implement, mainly, node-based hooks. This change should not alter their function - node-based hooks should run as before.

👍 

> 
> Offhand, do you know if the UI has any issues tracking out-of-process execution? I see that `HamiltonTracker` is stateful and I don't see any evident `__getstate__` or `__setstate__` methods, but I could be missing something. If there are issues, these hooks might be of some use - I would need to experiment.

Yes, that's in the client that sends the events.

elijahbenizzy commented:
> > I am not that familiar with the `HamiltonTracker`, but it appears to implement, mainly, node-based hooks. This change should not alter their function - node-based hooks should run as before.
> 
> 👍
> 
> > Offhand, do you know if the UI has any issues tracking out-of-process execution? I see that `HamiltonTracker` is stateful and I don't see any evident `__getstate__` or `__setstate__` methods, but I could be missing something. If there are issues, these hooks might be of some use - I would need to experiment.
> 
> Yes, that's in the client that sends the events - need to go one level deeper in the tracker to the client it uses.

To chime in -- I don't think this should ipmact the hamilton tracker. That said, we may want to consider adding more to it to leverage these -- we could get a more fine-grained view/get expose more descriptive states.

skrawcz commented:
Cool. Does it makes sense to provide some implementation of these hooks in this PR too? e.g. some example using it?

cswartzvi commented:
@skrawcz - I was actually planning to create a second PR with a dedicated logging adapter (based on a side conversation with @elijahbenizzy) using these new hooks. I can add that here if you'd like.

skrawcz commented:
> @skrawcz - I was actually planning to create a second PR with a dedicated logging adapter (based on a side conversation with @elijahbenizzy) using these new hooks. I can add that here if you'd like.

Cool. Yeah I say having an implementation would be good because having something implemented helps prove / ground the API :) It can be in a PR off of this branch or together.

cswartzvi commented:
So, the logging adapter is a bit more expansive than I originally thought. We will probably want to have a discussion. I will open a separate PR based off this branch - hopefully this evening.

elijahbenizzy commented on hamilton/execution/executors.py:
```diff
@@ -396,20 +407,32 @@ def run_graph_to_completion(
                             f"{[item.name for item in next_task.nodes]}"
                         )
                         raise e
-                    task_futures[next_task.task_id] = submitted
```
Why the change here? Assuming `Task` is hashable? Don't see it in the spec...

elijahbenizzy replied:
Oh, see you changed it... Got it, this allows us to have the extra data available...

cswartzvi replied:
Yup that was what I was going for! I needed to access the `adapater` for the task and thought adding a parallel structure was a little redundent. Do you see any problems with making `TaskImplementation` hashable?

elijahbenizzy commented on tests/lifecycle/test_lifecycle_adapters_end_to_end_task_based.py:
```diff
@@ -242,6 +246,72 @@ def test_individual_post_task_expand_hook():
+    spawning_task_ids = Counter([item.bound_kwargs["spawning_task_id"] for item in relevant_calls])
+    assert spawning_task_ids == {"expand-parallel_over": 5, None: 5}
+    purposes = Counter([item.bound_kwargs["purpose"] for item in relevant_calls])
+    assert purposes == {
```
Comments on this might be nice -- E.G. why these are the call counts

elijahbenizzy replied:
(although I realize that you probably copied this approach from others which don't document...)

cswartzvi replied:
Good idea! It took me a few minutes to re-figure out what the counts were looking at them now - a sure sign they are need. Adding now ...

elijahbenizzy reviewed (commented):
Looking good -- a few minor comments

cswartzvi commented:
Summary of the latest changes (I have been sitting on them)...

- Made `TaskFuture` a protocol, added a `TaskFutureWrappingFunction` implementation - this was done so that errors could be handled correctly from synchronous tasks
- Cleaned up some of the exception handling at submission and resolution calling site
- Various other quality of life improvements

cswartzvi commented:
Hmm ... tests are failing because of `xgboost`. Not sure if this was something I broke.

elijahbenizzy commented on hamilton/execution/executors.py:
```diff
@@ -162,6 +166,36 @@ def base_execute_task(task: TaskImplementation) -> Dict[str, Any]:
+
+    def get_state(self):
+        try:
+            self._results = self.function()
```
Shouldn't this only call the function once? E.G. it should be cached?

cswartzvi replied:
Yup! This was an oversight. Commit inbound.

elijahbenizzy commented on hamilton/execution/executors.py:
```diff
@@ -172,9 +206,7 @@ def submit_task(self, task: TaskImplementation) -> TaskFuture:
-        # No error management for now
-        result = base_execute_task(task)
-        return TaskFuture(get_state=lambda: TaskState.SUCCESSFUL, get_result=lambda: result)
+        return TaskFutureWrappingFunction(functools.partial(base_execute_task, task))
```
Does this add error management?

cswartzvi replied:
I believe that I removed that comment prematurely. There is no direct error management in `TaskFutureWrappingFunction`. However, there is now error management around the `post_task_resolution` hook in `run_graph_to_completion` (it gets sent to the hook before propagating upward). I would be happy to add the comment back if you like!

elijahbenizzy commented on hamilton/execution/executors.py:
```diff
@@ -383,11 +416,22 @@ def run_graph_to_completion(
+                            purpose=next_task.purpose,
+                        )
                     try:
                         submitted = task_executor.submit_task(next_task)
```
Not now but at some point a post_task_submission hook might make sense.

cswartzvi replied:
Yeah that is a good idea 💯

elijahbenizzy commented on hamilton/lifecycle/api.py:
```diff
@@ -371,10 +373,121 @@ def run_after_graph_execution(
+
+
+class TaskResolutionHook(BasePostTaskResolution, abc.ABC):
+    """Implement this to hook into the task resolution process. Tasks are submitted to an executor,
```
Let's clarify a little more what task "resolution" is for users -- it's nuanced and a bit confusing.

IIUC this is what happens when we have a task "template", which gets realized/ "bound" to tasks. E.G. when all the dependencies are satisfied and we can enqueue it. Right?

cswartzvi replied:
I struggled with the name of this hook/adapter. Currently the `post_task_resolution` hook gets called in `run_graph_to_completion` after the task future gets resolved. Maybe it's better to describe it as 'a hook that fires after a task returns from an executor'? Ha - that is actually a better complement to `pre_task_submission`! What do you think - I would happy to rename the hook/adapter - maybe `BasePostTaskReturn` / `TaskReturnnHook`?

elijahbenizzy replied:
Yeah I like `post_task_return` and `TaskReturnHooks` -- I think that's a bit clearer.

elijahbenizzy reviewed (commented):
Looks good! A few small changes but I think it's almost there.

elijahbenizzy reviewed (approved):
Thanks! Sorry this took a bit to review. Looks good, just name change the hook and ping me, I'll merge.

cswartzvi commented:
@elijahbenizzy I changed the name of the hook and merged in main (probably should have rebased).
